differs/gem_diff.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2017 Google, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differs
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	pkgutil "github.com/GoogleCloudPlatform/container-diff/pkg/util"
+	"github.com/GoogleCloudPlatform/container-diff/util"
+	"github.com/sirupsen/logrus"
+)
+
+type GemAnalyzer struct {
+}
+
+func (a GemAnalyzer) Name() string {
+	return "GemAnalyzer"
+}
+
+// GemAnalyzer.Diff compares gem/bundler installed Ruby packages between layers of two different images
+func (a GemAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := multiVersionDiff(image1, image2, a)
+	return diff, err
+}
+
+func (a GemAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := multiVersionAnalysis(image, a)
+	return analysis, err
+}
+
+func (a GemAnalyzer) getPackages(image pkgutil.Image) (map[string]map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	packages := make(map[string]map[string]util.PackageInfo)
+	rubyPaths := []string{}
+	if image.Config.Config.Env != nil {
+		paths := getGemPaths(image.Config.Config.Env)
+		for _, p := range paths {
+			rubyPaths = append(rubyPaths, p)
+		}
+	}
+	rubyVersions, err := getRubyVersion(path)
+	if err != nil {
+		logrus.Warnf("Image doesn't appear to have Ruby installed")
```

Comment:
looks like if we get here packages hasn't been added to, so it's fine to return, but go ahead and return the error here instead of logging it. the top level caller should be responsible for logging everything. do the same for the other errors below too.

```return packages, errors.New("Image....")```